[cutlass backend] Add serializer for cutlass ops

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153894

Differential Revision: [D74524786](https://our.internmc.facebook.com/intern/diff/D74524786/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

henrylhtsang committed:
[cutlass backend] Add serializer for cutlass ops



Differential Revision: [D74524786](https://our.internmc.facebook.com/intern/diff/D74524786/)

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] Add serializer for cutlass ops"


Differential Revision: [D74524786](https://our.internmc.facebook.com/intern/diff/D74524786/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ColinPeppler commented on torch/_inductor/codegen/cuda/serialization.py:
```diff
@@ -0,0 +1,464 @@
+        return cls._json_to_gemm_operation(json_dict)
+
+    @classmethod
+    def _gemm_operation_to_json(cls, operation):
```
looks clean, if you add types for inputs & outputs this would be much cleaner :)

henrylhtsang commented on test/inductor/test_cutlass_backend.py:
```diff
@@ -1585,6 +1610,31 @@ def forward(self, a, b, extra_args):
+                        count += 1
+                        serialized = serializer.serialize(op)
+                        deserialized = serializer.deserialize(serialized)
+                        _check_if_instances_equal(op, deserialized)
```
OMG how did I miss that

henrylhtsang committed:
Update on "[cutlass backend] Add serializer for cutlass ops"


Differential Revision: [D74524786](https://our.internmc.facebook.com/intern/diff/D74524786/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ColinPeppler commented on test/inductor/test_cutlass_backend.py:
```diff
@@ -63,6 +68,34 @@ def _get_path_without_sccache() -> str:
+    if op1.__class__.__dict__.get("__eq__") and op1 != op2:
+        return False
+
+    if isinstance(op1, Enum):
+        return op1.value == op2.value
```
this just checks the value? is it possible for op1 and op2 to be two separate unrelated enums and have the same value?

```
class EnumA(Enum):
  x = 1

class EnumB(Enum):
  y = 1

EnumA.x.value == EnumB.y.value # True
```

henrylhtsang replied:
there is a type(op1) == type(op2) check earlier

ColinPeppler commented on torch/_inductor/codegen/cuda/serialization.py:
```diff
@@ -0,0 +1,465 @@
+        return ret
+
+    @classmethod
+    def deserialize(cls, json_str: str) -> "GemmOperation":  # type: ignore[name-defined]  # noqa: F821
```
nit: you may be able to do smth like this to avoid the linter error

```
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from ... import ...  
```

ColinPeppler commented on torch/_inductor/codegen/cuda/serialization.py:
```diff
@@ -0,0 +1,465 @@
+            return None
+
+        # Create a dictionary for math_instruction if it exists
+        math_instruction_dict = None
+        if (
+            hasattr(tile_desc, "math_instruction")
+            and tile_desc.math_instruction is not None
+        ):
+            math_instruction = tile_desc.math_instruction
+            math_instruction_dict = {
+                "instruction_shape": math_instruction.instruction_shape,
+                "element_a": cls._enum_to_json(math_instruction.element_a),
+                "element_b": cls._enum_to_json(math_instruction.element_b),
+                "element_accumulator": cls._enum_to_json(
+                    math_instruction.element_accumulator
+                ),
+                "opcode_class": cls._enum_to_json(math_instruction.opcode_class),
+                "math_operation": cls._enum_to_json(math_instruction.math_operation),
+            }
```
I think it's fine for today, but at some point CUTLASS can add whatever new field to this class and it will break this.

I have a feeling it's better to iterate through the `vars(tile_desc.math_instruction)` and create key, values accordingly. But that can be a follow-up.

henrylhtsang commented:
@pytorchbot merge

henrylhtsang commented:
@pytorchbot merge -i

nWEIdia commented:
Hi @henrylhtsang , which cutlass version is this designed and tested with? 

Encountering AttributeError: 'TileDescription' object has no attribute 'explicit_vector_sizes'
 
The pytorch source code is 3.9.2, but the currently available PYPI package is 3.8.0. Is the above explicit_vector_sizes coming from the difference between 3.8.0 cutlass and 3.9.2? 

cc @ptrblck @eqy @Aidyn-A

henrylhtsang commented:
> Hi @henrylhtsang , which cutlass version is this designed and tested with?
> 
> Encountering AttributeError: 'TileDescription' object has no attribute 'explicit_vector_sizes'
> 
> The pytorch source code is 3.9.2, but the currently available PYPI package is 3.8.0. Is the above explicit_vector_sizes coming from the difference between 3.8.0 cutlass and 3.9.2?
> 
> cc @ptrblck @eqy @Aidyn-A

3.9.2. What is your cutlass version

nWEIdia commented:
> > Hi @henrylhtsang , which cutlass version is this designed and tested with?
> > Encountering AttributeError: 'TileDescription' object has no attribute 'explicit_vector_sizes'
> > The pytorch source code is 3.9.2, but the currently available PYPI package is 3.8.0. Is the above explicit_vector_sizes coming from the difference between 3.8.0 cutlass and 3.9.2?
> > cc @ptrblck @eqy @Aidyn-A
> 
> 3.9.2. What is your cutlass version
> 
> also curious: how did you bump into this

Mainly because I used "pip install nvidia-cutlass" which has only 3.8.0 in PYPI https://pypi.org/project/nvidia-cutlass/ . 
If I do not use "pip install nvidia-cutlass", then I am encountering 

No module named 'torch._inductor.codegen.cuda.cutlass_lib_extensions'`   with  test/inductor/test_cutlass_backend.py::TestCutlassBackend::test_evt_fusions_basic_sub_shape3    (using /pytorch/third_party/cutlass)

henrylhtsang commented:
> > > > > Hi @henrylhtsang , which cutlass version is this designed and tested with?
> > > > > Encountering AttributeError: 'TileDescription' object has no attribute 'explicit_vector_sizes'
> > > > > The pytorch source code is 3.9.2, but the currently available PYPI package is 3.8.0. Is the above explicit_vector_sizes coming from the difference between 3.8.0 cutlass and 3.9.2?
> > > > > cc @ptrblck @eqy @Aidyn-A
> > > > 
> > > > 
> > > > 3.9.2. What is your cutlass version
> > > > also curious: how did you bump into this
> > > 
> > > 
> > > Mainly because I used "pip install nvidia-cutlass" which has only 3.8.0 in PYPI https://pypi.org/project/nvidia-cutlass/ . If I do not use "pip install nvidia-cutlass", then I am encountering
> > > No module named 'torch._inductor.codegen.cuda.cutlass_lib_extensions'` with test/inductor/test_cutlass_backend.py::TestCutlassBackend::test_evt_fusions_basic_sub_shape3 (using /pytorch/third_party/cutlass)
> > 
> > 
> > How are you using pytorch right now? do you build from source?
> > umm I feel like like pytorch/third_party/cutlass should work
> 
> Yes I am using pytorch that was built from source.
> 
> Using pytorch/third_party/cutlass, the following unit test passes (without using nvidia-cutlass 3.8.0): pytest -v test/inductor/test_cutlass_backend.py -k test_gemm_operation_serialization_arch_100_cuda_version_12_8
> 
> but then encountering `No module named 'torch._inductor.codegen.cuda.cutlass_lib_extensions'` with test/inductor/test_cutlass_backend.py::TestCutlassBackend::test_evt_fusions_basic_sub_shape3
> 
> So I went back to "pip install nvidia-cutlass", then pytest -v test/inductor/test_cutlass_backend.py -k test_gemm_operation_serialization_arch_100_cuda_version_12_8 would encounter the
> 
> FAILED [0.4198s] test/inductor/test_cutlass_backend.py::TestCutlassBackend::test_gemm_operation_serialization_arch_100_cuda_version_12_8 - AttributeError: 'TileDescription' object has no attribute 'explicit_vector_sizes'
> 
> Did you build cutlass wheel (3.9.2) from src? Or you were using the third_party/cutlass source code only?

Let me try to reproduce first. You should be able to run the tests fine without `pip install nvidia-cutlass`

Did you try a clean install? `python setup.py clean`

nWEIdia commented:
Could you please reproduce it using nightly docker image? 
1. Get a x86 based H100 (because there is no arm version of the docker image and  on A100  the test would skip) 
2. docker pull ghcr.io/pytorch/pytorch-nightly:2.8.0.dev20250612-cuda12.8-cudnn9-devel 
3. (make sure nvidia-container-toolkit is installed, if not, please follow the install instructions https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html 
4. docker run --gpus all -it ghcr.io/pytorch/pytorch-nightly:2.8.0.dev20250612-cuda12.8-cudnn9-devel  bash 
5. apt install git -y && git clone --single-branch --branch main https://github.com/pytorch/pytorch.git
6.  pip install pytest
7.   pytest -v  test/inductor/test_cutlass_backend.py::TestCutlassBackend::test_evt_fusions_basic_sub_shape3

henrylhtsang commented:
@nWEIdia I think the quickest way to unblock is to make the serialization compatible with cutlass 3.8. Feel free to do it, since I don't see much value in it especially when nvidia stops updating the nvidia-cutlass pypi package.

Another way to do it is to just forget about the serialization logic. What is your use case?

henrylhtsang commented:
> For your information, one of the things that was needed to unblock was to hard code pytorch source code location,

can you elaborate on "For your information, one of the things that was needed to unblock was to hard code pytorch source code location"

> I am following-up on nvidia-cutlass not publishing 3.9+ as well.

By the way, since you seem to have leverage there, is it possible to request a name change for the python/cutlass? since now the name is in conflict with nvidia-cutlass-dsl...

nWEIdia commented:
> can you elaborate on "For your information, one of the things that was needed to unblock was to hard code pytorch source code location"

Perhaps I was missing some setup. So let's assume nvidia-cutlass is not available, as non-metamates, the only option is to use ../third_party/cutlass, as commented in https://github.com/pytorch/pytorch/pull/145891 
As a user, the only thing that comes to my mind is PYTORCH_SRC_ROOT_DIRECTORY/third_party/cutlass. However the test_import_cutlass function in test_cutlass_backend.py is using "config.cuda.cutlass_dir"  (is this guranteed to be the src pytorch directory/third_party/cutlass?). Or is the expectation that pytorch wheel **Installtion** directory would also have third_party/cutlass? (I don't think so) .

> > I am following-up on nvidia-cutlass not publishing 3.9+ as well.
> 
> By the way, since you seem to have leverage there, is it possible to request a name change for the python/cutlass? since now the name is in conflict with nvidia-cutlass-dsl...

cc @ jackkosaian , could you please provide more details regarding "nvidia-cutlass-dsl" vs "python/cutlass" I do not see a connection here. 

(p.s. would be OOTO for ~10 days)

nWEIdia commented:
> > > can you elaborate on "For your information, one of the things that was needed to unblock was to hard code pytorch source code location"
> > 
> > 
> > Perhaps I was missing some setup. So let's assume nvidia-cutlass is not available, as non-metamates, the only option is to use ../third_party/cutlass, as commented in #145891 As a user, the only thing that comes to my mind is PYTORCH_SRC_ROOT_DIRECTORY/third_party/cutlass. However the test_import_cutlass function in test_cutlass_backend.py is using "config.cuda.cutlass_dir" (is this guranteed to be the src pytorch directory/third_party/cutlass?). Or is the expectation that pytorch wheel **Installtion** directory would also have third_party/cutlass? (I don't think so) . **Do I need a step to compile cutlass source before running test_cutlass_backend?**
> > > > I am following-up on nvidia-cutlass not publishing 3.9+ as well.
> > > 
> > > 
> > > By the way, since you seem to have leverage there, is it possible to request a name change for the python/cutlass? since now the name is in conflict with nvidia-cutlass-dsl...
> > 
> > 
> > cc @ jackkosaian , could you please provide more details regarding "nvidia-cutlass-dsl" vs "python/cutlass" I do not see a connection here.
> > (p.s. would be OOTO for ~10 days)
> 
> I think the way is to specify TORCHINDUCTOR_CUTLASS_DIR to where cutlass dir is located.
> 
> Basically, say you use pytorch wheel. It won't come with cutlass. Forgetting the now outdated nvidia-cutlass, the only option is to clone github cutlass, and specify where it is located with TORCHINDUCTOR_CUTLASS_DIR. Though admitted I didn't test it.
> 
> No you don't need to build cutlass.

Thank you for the clarification! Yeah, currently our internal CI workflow does not support specifying where cutlass is located with TORCHINDUCTOR_CUTLASS_DIR, I will try manually with this option and see. 

BTW, to clarify, the question is for you "could you please provide more details regarding "nvidia-cutlass-dsl" vs "python/cutlass" I do not see a connection here."

henrylhtsang commented:
> > > > can you elaborate on "For your information, one of the things that was needed to unblock was to hard code pytorch source code location"
> > > 
> > > 
> > > Perhaps I was missing some setup. So let's assume nvidia-cutlass is not available, as non-metamates, the only option is to use ../third_party/cutlass, as commented in #145891 As a user, the only thing that comes to my mind is PYTORCH_SRC_ROOT_DIRECTORY/third_party/cutlass. However the test_import_cutlass function in test_cutlass_backend.py is using "config.cuda.cutlass_dir" (is this guranteed to be the src pytorch directory/third_party/cutlass?). Or is the expectation that pytorch wheel **Installtion** directory would also have third_party/cutlass? (I don't think so) . **Do I need a step to compile cutlass source before running test_cutlass_backend?**
> > > > > I am following-up on nvidia-cutlass not publishing 3.9+ as well.
> > > > 
> > > > 
> > > > By the way, since you seem to have leverage there, is it possible to request a name change for the python/cutlass? since now the name is in conflict with nvidia-cutlass-dsl...
> > > 
> > > 
> > > cc @ jackkosaian , could you please provide more details regarding "nvidia-cutlass-dsl" vs "python/cutlass" I do not see a connection here.
> > > (p.s. would be OOTO for ~10 days)
> > 
> > 
> > I think the way is to specify TORCHINDUCTOR_CUTLASS_DIR to where cutlass dir is located.
> > Basically, say you use pytorch wheel. It won't come with cutlass. Forgetting the now outdated nvidia-cutlass, the only option is to clone github cutlass, and specify where it is located with TORCHINDUCTOR_CUTLASS_DIR. Though admitted I didn't test it.
> > No you don't need to build cutlass.
> 
> Thank you for the clarification! Yeah, currently our internal CI workflow does not support specifying where cutlass is located with TORCHINDUCTOR_CUTLASS_DIR, I will try manually with this option and see.
> 
> BTW, to clarify, the question is for you "could you please provide more details regarding "nvidia-cutlass-dsl" vs "python/cutlass" I do not see a connection here."

potentially this is a noob question, so bare with me.

We currently have use cases for nvidia-cutlass. More precisely, it is the python/cutlass for EVT. So in our use cases, we would do `import cutlass`.

Now, with `nvidia-cutlass-dsl` coming up, apparently the ui is also `import cutlass`. This is a problem because python doesn't support two modules having the same name. So this would mean we won't be able to use both at the same time, at least without some hacking. So my question is, will there be a solution on this? I heard @mlazos talked about in the future you guys will move the EVT part to python/cutlass_library. Any ETA for that?

nWEIdia commented:
> > > > > can you elaborate on "For your information, one of the things that was needed to unblock was to hard code pytorch source code location"
> > > > 
> > > > 
> > > > Perhaps I was missing some setup. So let's assume nvidia-cutlass is not available, as non-metamates, the only option is to use ../third_party/cutlass, as commented in #145891 As a user, the only thing that comes to my mind is PYTORCH_SRC_ROOT_DIRECTORY/third_party/cutlass. However the test_import_cutlass function in test_cutlass_backend.py is using "config.cuda.cutlass_dir" (is this guranteed to be the src pytorch directory/third_party/cutlass?). Or is the expectation that pytorch wheel **Installtion** directory would also have third_party/cutlass? (I don't think so) . **Do I need a step to compile cutlass source before running test_cutlass_backend?**
> > > > > > I am following-up on nvidia-cutlass not publishing 3.9+ as well.
> > > > > 
> > > > > 
> > > > > By the way, since you seem to have leverage there, is it possible to request a name change for the python/cutlass? since now the name is in conflict with nvidia-cutlass-dsl...
> > > > 
> > > > 
> > > > cc @ jackkosaian , could you please provide more details regarding "nvidia-cutlass-dsl" vs "python/cutlass" I do not see a connection here.
> > > > (p.s. would be OOTO for ~10 days)
> > > 
> > > 
> > > I think the way is to specify TORCHINDUCTOR_CUTLASS_DIR to where cutlass dir is located.
> > > Basically, say you use pytorch wheel. It won't come with cutlass. Forgetting the now outdated nvidia-cutlass, the only option is to clone github cutlass, and specify where it is located with TORCHINDUCTOR_CUTLASS_DIR. Though admitted I didn't test it.
> > > No you don't need to build cutlass.
> > 
> > 
> > Thank you for the clarification! Yeah, currently our internal CI workflow does not support specifying where cutlass is located with TORCHINDUCTOR_CUTLASS_DIR, I will try manually with this option and see.
> > BTW, to clarify, the question is for you "could you please provide more details regarding "nvidia-cutlass-dsl" vs "python/cutlass" I do not see a connection here."
> 
> potentially this is a noob question, so bare with me.
> 
> We currently have use cases for nvidia-cutlass. More precisely, it is the python/cutlass for EVT. So in our use cases, we would do `import cutlass`.
> 
> Now, with `nvidia-cutlass-dsl` coming up, apparently the ui is also `import cutlass`. This is a problem because python doesn't support two modules having the same name. So this would mean we won't be able to use both at the same time, at least without some hacking. So my question is, will there be a solution on this? I heard @mlazos talked about in the future you guys will move the EVT part to python/cutlass_library. Any ETA for that?

Thanks for the additional details! 
Now, cc @jackkosaian and @eqy to help follow-up on this.  (as I do not know the answer).
